### PR TITLE
Ivy publish should retry publication on transient remote errors 

### PR DESCRIPTION
**Pull request by rpalcolea**
### Context
Similar to https://github.com/gradle/gradle/issues/8992, Ivy publishing should support retries.

I took the original implementation for maven: https://github.com/gradle/gradle/pull/9163 and extracted `MavenDeployRetrier` and `DefaultMavenDeployRetrier`  into a generic retrier and then applied to `PublishToIvyRepository` task used/created by `IvyPublishPlugin`

### Contributor Checklist
- [x] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md)
- [x] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#git-commit---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [x] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [x] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [x] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [x] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [x] Ensure that tests pass locally: `./gradlew <changed-subproject>:check`

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation
- [ ] Recognize contributor in release notes

### CODE REVIEW

File: subprojects/publish/src/main/java/org/gradle/api/publish/PublishRetrier.java [hunk 1]
```diff
@@ -14,10 +14,22 @@
...
 
-interface MavenDeployRetrier {
-    void deployWithRetry() throws DeploymentException;
+/**
+ * Publisher that is able to retry in case of intermittent network issues.
+ *
+ * @since 6.3
+ */
+@Incubating
+public interface PublishRetrier {
```
Comment (melix) [inline subprojects/publish/src/main/java/org/gradle/api/publish/PublishRetrier.java]: :x: This shouldn't be public API IMHO. Any reason why you put this on the public API?

File: subprojects/ivy/src/main/java/org/gradle/api/publish/ivy/tasks/PublishToIvyRepository.java [hunk 2]
```diff
@@ -40,8 +41,13 @@
  */
 public class PublishToIvyRepository extends DefaultTask {
 
+    private final static String MAX_DEPLOY_ATTEMPTS = "org.gradle.internal.remote.repository.deploy.max.attempts";
```
Comment (melix) [inline subprojects/ivy/src/main/java/org/gradle/api/publish/ivy/tasks/PublishToIvyRepository.java]: :x: Those constants should be shared between the Ivy and Maven implementation.

File: subprojects/publish/src/main/java/org/gradle/api/publish/PublishRetrierException.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+import org.gradle.api.Incubating;
+
+/**
+ *
+ * Exception used for errors associated with {@link PublishRetrier}
+ * @since 6.3
+ */
+@Incubating
+public class PublishRetrierException extends Exception {
```
Comment (melix) [inline subprojects/publish/src/main/java/org/gradle/api/publish/PublishRetrierException.java]: :x:Same comment here, I don't think we should make this public API.

Comment (rpalcolea) [reply subprojects/publish/src/main/java/org/gradle/api/publish/PublishRetrierException.java]: Addressed in https://github.com/gradle/gradle/pull/12024/commits/700fac49f84913d6f73d6b87226c6645da89c029

File: subprojects/ivy/src/main/java/org/gradle/api/publish/ivy/tasks/PublishToIvyRepository.java (hunk 2 above)
Comment (rpalcolea) [reply subprojects/ivy/src/main/java/org/gradle/api/publish/ivy/tasks/PublishToIvyRepository.java]: Addressed https://github.com/gradle/gradle/pull/12024/commits/11b90ffe4936d0907ae054a56080ec4b01585578

File: subprojects/publish/src/main/java/org/gradle/api/publish/PublishRetrier.java (hunk 1 above)
Comment (rpalcolea) [reply subprojects/publish/src/main/java/org/gradle/api/publish/PublishRetrier.java]: Addressed in https://github.com/gradle/gradle/pull/12024/commits/700fac49f84913d6f73d6b87226c6645da89c029

**Comment by rpalcolea** [comment]
Hi @melix, thanks for the feedback.

I switched to internal APIs and also re-using the flags for initial values.

The current tests cover the `PublisherRetrier` and https://github.com/gradle/gradle/blob/master/subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/repositories/transport/NetworkingIssueVerifierTest.groovy to make sure that if a known exception is bubbled up by gradle, we mark that as something to retry.

With that said, there is nothing to exercise https://github.com/gradle/gradle/pull/12024/files#diff-07d3fd0a840441700a2e2d9c1fa58e38R152. So basically blind trusting that Gradle will bubble up exceptions when network issues happen.

I took a quick look at the code base and didn't seem to find any opinion around mocking a repository using something like wiremock or creating `HttpServer` to respond with a 503 for example which would trigger the behavior. I feel that this would add overhead to the test suite. Are there any opinions or existing guidance or adding tests like this to the integration test suite? thanks!

**Comment by rpalcolea** [comment]
Hey @melix ,

I found that @bigdaz already added this in:

https://github.com/gradle/gradle/blob/master/subprojects/ivy/src/main/java/org/gradle/api/publish/ivy/internal/publisher/DependencyResolverIvyPublisher.java#L48
https://github.com/gradle/gradle/blob/master/subprojects/maven/src/main/java/org/gradle/api/publish/maven/internal/publisher/AbstractMavenPublisher.java#L252

So opened https://github.com/gradle/gradle/pull/12127 with some tests and removed extra cruft added by one of my PRs last year: https://github.com/gradle/gradle/pull/9163

**Pull request closed without merging**